File: src/main/metadados.js
'use strict';
/**
 * Organizacao da biblioteca: pastas, capas, nomes de exibicao, descricoes e
 * etiquetas.
 *
 * Fica num arquivo proprio (organizacao.json), separado do catalogo. O catalogo
 * e recontruido a cada ciclo a partir do qBittorrent; o que voce edita aqui nao
 * pode ser atropelado por isso.
 *
 * As pastas sao VIRTUAIS: existem so dentro do app. Renomear, mover ou apagar
 * uma pasta nunca toca num arquivo em disco nem no torrent que o alimenta.
 */
const { app, net } = require('electron');
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const ARQUIVO = () => path.join(app.getPath('userData'), 'organizacao.json');
const PASTA_CAPAS = () => path.join(app.getPath('userData'), 'capas');

const EXT_IMAGEM = new Set(['.jpg', '.jpeg', '.png', '.webp', '.gif', '.avif', '.bmp']);
const TAMANHO_MAXIMO_CAPA = 12 * 1024 * 1024; // 12 MB

const MIMES = {
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.png': 'image/png',
    '.webp': 'image/webp',
    '.gif': 'image/gif',
    '.avif': 'image/avif',
    '.bmp': 'image/bmp',
};

let dados = null;

// --------------------------------------------------------------- persistencia

function vazio() {
    return { pastas: {}, titulos: {} };
}

function carregar() {
    if (dados) return dados;
    try {
        const lido = JSON.parse(fs.readFileSync(ARQUIVO(), 'utf8'));
        dados = Object.assign(vazio(), lido);
        dados.pastas = dados.pastas || {};
        dados.titulos = dados.titulos || {};
    } catch {
        dados = vazio();
    }
    return dados;
}

let pendente = null;
function gravar() {
    clearTimeout(pendente);
    pendente = setTimeout(() => {
        try {
            fs.mkdirSync(path.dirname(ARQUIVO()), { recursive: true });
            fs.writeFileSync(ARQUIVO(), JSON.stringify(dados, null, 2), 'utf8');
        } catch (erro) {
            console.error('Falha ao gravar a organizacao:', erro.message);
        }
    }, 300);
}

function novoId() {
    return crypto.randomBytes(8).toString('hex');
}

function texto(valor, limite = 4000) {
    return typeof valor === 'string' ? valor.trim().slice(0, limite) : '';
}

// -------------------------------------------------------------------- pastas

/** Impede que uma pasta vire descendente de si mesma ao ser movida. */
function ehDescendente(idPossivelFilho, idPossivelPai) {
    let atual = dados.pastas[idPossivelFilho];
    const vistos = new Set();
    while (atual && atual.pai && !vistos.has(atual.pai)) {
        if (atual.pai === idPossivelPai) return true;
        vistos.add(atual.pai);
        atual = dados.pastas[atual.pai];
    }
    return false;
}

function criarPasta({ nome, pai = null }) {
    carregar();
    const id = novoId();
    dados.pastas[id] = {
        id,
        nome: texto(nome, 120) || 'Nova pasta',
        descricao: '',
        capa: null,
        pai: pai && dados.pastas[pai] ? pai : null,
        criadaEm: Date.now(),
    };
    gravar();
    return dados.pastas[id];
}

function editarPasta(id, campos) {
    carregar();
    const pasta = dados.pastas[id];
    if (!pasta) return null;

    if (campos.nome !== undefined) pasta.nome = texto(campos.nome, 120) || pasta.nome;
    if (campos.descricao !== undefined) pasta.descricao = texto(campos.descricao);
    if (campos.pai !== undefined) {
        const alvo = campos.pai;
        const valido = alvo === null || (dados.pastas[alvo] && alvo !== id && !ehDescendente(alvo, id));
        if (valido) pasta.pai = alvo || null;
    }
    gravar();
    return pasta;
}

/** Remove a pasta. O que estava dentro sobe um nivel -- nada se perde. */
function removerPasta(id) {
    carregar();
    const pasta = dados.pastas[id];
    if (!pasta) return false;

    const destino = pasta.pai || null;
    for (const outra of Object.values(dados.pastas)) {
        if (outra.pai === id) outra.pai = destino;
    }
    for (const titulo of Object.values(dados.titulos)) {
        if (titulo.pasta === id) titulo.pasta = destino;
    }
    apagarArquivoCapa(pasta.capa);
    delete dados.pastas[id];
    gravar();
    return true;
}

function listarPastas() {
    carregar();
    return Object.values(dados.pastas).sort((a, b) => a.nome.localeCompare(b.nome, 'pt-BR'));
}

/** Caminho da raiz ate a pasta, para a trilha de navegacao. */
function trilha(id) {
    carregar();
    const caminho = [];
    const vistos = new Set();
    let atual = id ? dados.pastas[id] : null;
    while (atual && !vistos.has(atual.id)) {
        caminho.unshift({ id: atual.id, nome: atual.nome });
        vistos.add(atual.id);
        atual = atual.pai ? dados.pastas[atual.pai] : null;
    }
    return caminho;
}

// ------------------------------------------------------------------ titulos

function doTitulo(hash) {
    carregar();
    if (!dados.titulos[hash]) {
        dados.titulos[hash] = { nome: '', descricao: '', capa: null, etiquetas: [], pasta: null, arquivos: {} };
    }
    return dados.titulos[hash];
}

function editarTitulo(hash, campos) {
    const t = doTitulo(hash);
    if (campos.nome !== undefined) t.nome = texto(campos.nome, 200);
    if (campos.descricao !== undefined) t.descricao = texto(campos.descricao);
    if (campos.etiquetas !== undefined) {
        const lista = Array.isArray(campos.etiquetas)
            ? campos.etiquetas
            : String(campos.etiquetas || '').split(',');
        t.etiquetas = [...new Set(lista.map((e) => texto(e, 40)).filter(Boolean))].slice(0, 20);
    }
    if (campos.pasta !== undefined) {
        t.pasta = campos.pasta && dados.pastas[campos.pasta] ? campos.pasta : null;
    }
    gravar();
    return t;
}

/** Nome de exibicao de um arquivo (episodio) dentro de um titulo. */
function editarArquivo(hash, caminho, nome) {
    const t = doTitulo(hash);
    t.arquivos = t.arquivos || {};
    const limpo = texto(nome, 200);
    if (limpo) t.arquivos[caminho] = { nome: limpo };
    else delete t.arquivos[caminho];
    gravar();
    return t;
}

// -------------------------------------------------------------------- capas

function garantirPastaCapas() {
    fs.mkdirSync(PASTA_CAPAS(), { recursive: true });
    return PASTA_CAPAS();
}

function apagarArquivoCapa(nome) {
    if (!nome) return;
    try {
        fs.unlinkSync(path.join(PASTA_CAPAS(), path.basename(nome)));
    } catch {
        /* ja nao existe */
    }
}

function extensaoValida(origem) {
    const ext = path.extname(new URL(origem, 'file:///').pathname).toLowerCase();
    return EXT_IMAGEM.has(ext) ? ext : null;
}

async function baixarImagem(url) {
    const resposta = await net.fetch(url);
    if (!resposta.ok) throw new Error(`o servidor respondeu ${resposta.status}`);

    const tipo = resposta.headers.get('content-type') || '';
    if (!tipo.startsWith('image/')) throw new Error(`o link nao aponta para uma imagem (${tipo || 'sem tipo'})`);

    const buffer = Buffer.from(await resposta.arrayBuffer());
    if (buffer.length > TAMANHO_MAXIMO_CAPA) throw new Error('a imagem passa de 12 MB');

    const porMime = Object.entries(MIMES).find(([, m]) => m === tipo.split(';')[0].trim());
    const ext = extensaoValida(url) || (porMime ? porMime[0] : '.jpg');
    return { buffer, ext };
}

function lerImagemLocal(caminho) {
    const ext = extensaoValida(caminho);
    if (!ext) throw new Error('formato de imagem nao suportado');
    const info = fs.statSync(caminho);
    if (info.size > TAMANHO_MAXIMO_CAPA) throw new Error('a imagem passa de 12 MB');
    return { buffer: fs.readFileSync(caminho), ext };
}

/**
 * Guarda a capa dentro dos dados do app -- uma COPIA, para a biblioteca nao
 * quebrar se o arquivo original for movido ou apagado depois.
 */
async function definirCapa({ tipo, id }, origem) {
    carregar();
    const alvo = tipo === 'pasta' ? dados.pastas[id] : doTitulo(id);
    if (!alvo) throw new Error('item nao encontrado');

    const { buffer, ext } = origem.url
        ? await baixarImagem(origem.url)
        : lerImagemLocal(origem.arquivo);

    garantirPastaCapas();
    const nome = `${novoId()}${ext}`;
    fs.writeFileSync(path.join(PASTA_CAPAS(), nome), buffer);

    apagarArquivoCapa(alvo.capa);
    alvo.capa = nome;
    gravar();
    return nome;
}

function removerCapa({ tipo, id }) {
    carregar();
    const alvo = tipo === 'pasta' ? dados.pastas[id] : dados.titulos[id];
    if (!alvo) return false;
    apagarArquivoCapa(alvo.capa);
    alvo.capa = null;
    gravar();
    return true;
}

/** Serve o arquivo da capa para o protocolo capa:// registrado no main. */
function lerCapa(nome) {
    const seguro = path.basename(String(nome || ''));
    const caminho = path.join(PASTA_CAPAS(), seguro);
    if (!seguro || !fs.existsSync(caminho)) return null;
    return { buffer: fs.readFileSync(caminho), mime: MIMES[path.extname(seguro).toLowerCase()] || 'image/jpeg' };
}

function urlDaCapa(nome) {
    return nome ? `capa://img/${encodeURIComponent(nome)}` : null;
}

// ------------------------------------------------------------------- juncao

/** Junta o catalogo da biblioteca com o que o usuario editou. */
function aplicar(entradas) {
    carregar();
    return entradas.map((e) => {
        const meta = dados.titulos[e.hash];
        if (!meta) return { ...e, nomeOriginal: e.nome, etiquetas: [], pasta: null, capa: null };

        return {
            ...e,
            nomeOriginal: e.nome,
            nome: meta.nome || e.nome,
            descricao: meta.descricao || '',
            etiquetas: meta.etiquetas || [],
            pasta: meta.pasta || null,
            capa: urlDaCapa(meta.capa),
            arquivos: e.arquivos.map((a) => ({
                ...a,
                nomeOriginal: a.nome,
                nome: (meta.arquivos && meta.arquivos[a.caminho] && meta.arquivos[a.caminho].nome) || a.nome,
            })),
        };
    });
}

function pastasParaInterface() {
    return listarPastas().map((p) => ({ ...p, capa: urlDaCapa(p.capa) }));
}

module.exports = {
    criarPasta,
    editarPasta,
    removerPasta,
    listarPastas,
    pastasParaInterface,
    trilha,
    editarTitulo,
    editarArquivo,
    definirCapa,
    removerCapa,
    lerCapa,
    urlDaCapa,
    aplicar,
};

File: android/app/src/main/java/com/torrange/app/Metadados.kt
package com.torrange.app

import android.content.Context
import android.net.Uri
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.security.SecureRandom
import java.util.concurrent.TimeUnit

/**
 * Organizacao da biblioteca: pastas, capas, nomes de exibicao, descricoes e
 * etiquetas.
 *
 * Fica num arquivo proprio (organizacao.json), separado do catalogo. O catalogo
 * e reconstruido a cada ciclo a partir do motor de torrent; o que voce edita
 * aqui nao pode ser atropelado por isso.
 *
 * As pastas sao VIRTUAIS: existem so dentro do aplicativo. Renomear, mover ou
 * apagar uma pasta nunca toca num arquivo em disco nem no torrent que o
 * alimenta -- entao nada quebra o seeding.
 */
class Metadados(private val contexto: Context) {

    companion object {
        private val EXT_IMAGEM = setOf("jpg", "jpeg", "png", "webp", "gif", "avif", "bmp")
        private const val TAMANHO_MAXIMO_CAPA = 12 * 1024 * 1024 // 12 MB

        private val MIMES = mapOf(
            "jpg" to "image/jpeg", "jpeg" to "image/jpeg", "png" to "image/png",
            "webp" to "image/webp", "gif" to "image/gif", "avif" to "image/avif",
            "bmp" to "image/bmp"
        )
    }

    private var dados: JSONObject? = null
    private var pendente = false

    private val cliente = OkHttpClient.Builder()
        .callTimeout(30, TimeUnit.SECONDS)
        .build()

    // ------------------------------------------------------------ persistencia

    @Synchronized
    private fun carregar(): JSONObject {
        dados?.let { return it }
        val lido = try {
            JSONObject(Caminhos.organizacao(contexto).readText())
        } catch (e: Exception) {
            JSONObject()
        }
        if (!lido.has("pastas")) lido.put("pastas", JSONObject())
        if (!lido.has("titulos")) lido.put("titulos", JSONObject())
        dados = lido
        return lido
    }

    private fun pastas(): JSONObject = carregar().getJSONObject("pastas")
    private fun titulos(): JSONObject = carregar().getJSONObject("titulos")

    @Synchronized
    private fun gravar() {
        if (pendente) return
        pendente = true
        Thread {
            Thread.sleep(300)
            synchronized(this) {
                pendente = false
                try {
                    Caminhos.organizacao(contexto).writeText(carregar().toString(2))
                } catch (e: Exception) {
                    Diagnostico.anotar("erro", "falha ao gravar a organizacao: ${e.message}")
                }
            }
        }.start()
    }

    private fun novoId(): String {
        val bytes = ByteArray(8)
        SecureRandom().nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it) }
    }

    private fun texto(valor: Any?, limite: Int = 4000): String =
        (valor as? String)?.trim()?.take(limite) ?: ""

    // ------------------------------------------------------------------ pastas

    /** Impede que uma pasta vire descendente de si mesma ao ser movida. */
    private fun ehDescendente(possivelFilho: String, possivelPai: String): Boolean {
        var atual = pastas().optJSONObject(possivelFilho)
        val vistos = HashSet<String>()
        while (atual != null) {
            val pai = atual.optString("pai").ifEmpty { null } ?: return false
            if (pai == possivelPai) return true
            if (!vistos.add(pai)) return false
            atual = pastas().optJSONObject(pai)
        }
        return false
    }

    @Synchronized
    fun criarPasta(dadosEntrada: JSONObject): JSONObject {
        val id = novoId()
        val pai = dadosEntrada.optString("pai").ifEmpty { null }
        val pasta = JSONObject()
            .put("id", id)
            .put("nome", texto(dadosEntrada.opt("nome"), 120).ifEmpty { "Nova pasta" })
            .put("descricao", "")
            .put("capa", JSONObject.NULL)
            .put("pai", if (pai != null && pastas().has(pai)) pai else JSONObject.NULL)
            .put("criadaEm", System.currentTimeMillis())
        pastas().put(id, pasta)
        gravar()
        return pasta
    }

    @Synchronized
    fun editarPasta(id: String, campos: JSONObject): JSONObject? {
        val pasta = pastas().optJSONObject(id) ?: return null

        if (campos.has("nome")) {
            val novo = texto(campos.opt("nome"), 120)
            if (novo.isNotEmpty()) pasta.put("nome", novo)
        }
        if (campos.has("descricao")) pasta.put("descricao", texto(campos.opt("descricao")))
        if (campos.has("pai")) {
            val alvo = campos.optString("pai").ifEmpty { null }
            val valido = alvo == null ||
                (pastas().has(alvo) && alvo != id && !ehDescendente(alvo, id))
            if (valido) pasta.put("pai", alvo ?: JSONObject.NULL)
        }
        gravar()
        return pasta
    }

    /** Remove a pasta. O que estava dentro sobe um nivel -- nada se perde. */
    @Synchronized
    fun removerPasta(id: String): Boolean {
        val pasta = pastas().optJSONObject(id) ?: return false
        val destino: Any = pasta.opt("pai") ?: JSONObject.NULL

        for (chave in pastas().keys()) {
            val outra = pastas().optJSONObject(chave) ?: continue
            if (outra.optString("pai") == id) outra.put("pai", destino)
        }
        for (chave in titulos().keys()) {
            val titulo = titulos().optJSONObject(chave) ?: continue
            if (titulo.optString("pasta") == id) titulo.put("pasta", destino)
        }
        apagarArquivoCapa(pasta.optString("capa"))
        pastas().remove(id)
        gravar()
        return true
    }

    @Synchronized
    fun listarPastas(): List<JSONObject> =
        pastas().keys().asSequence()
            .mapNotNull { pastas().optJSONObject(it) }
            .sortedBy { it.optString("nome").lowercase() }
            .toList()

    /** Caminho da raiz ate a pasta, para a trilha de navegacao. */
    @Synchronized
    fun trilha(id: String?): JSONArray {
        val caminho = mutableListOf<JSONObject>()
        val vistos = HashSet<String>()
        var atual = if (id.isNullOrEmpty()) null else pastas().optJSONObject(id)
        while (atual != null && vistos.add(atual.optString("id"))) {
            caminho.add(0, JSONObject().put("id", atual.optString("id")).put("nome", atual.optString("nome")))
            val pai = atual.optString("pai").ifEmpty { null }
            atual = if (pai != null) pastas().optJSONObject(pai) else null
        }
        return JSONArray().apply { caminho.forEach { put(it) } }
    }

    // ----------------------------------------------------------------- titulos

    @Synchronized
    private fun doTitulo(hash: String): JSONObject {
        titulos().optJSONObject(hash)?.let { return it }
        val novo = JSONObject()
            .put("nome", "")
            .put("descricao", "")
            .put("capa", JSONObject.NULL)
            .put("etiquetas", JSONArray())
            .put("pasta", JSONObject.NULL)
            .put("arquivos", JSONObject())
        titulos().put(hash, novo)
        return novo
    }

    @Synchronized
    fun editarTitulo(hash: String, campos: JSONObject): JSONObject {
        val t = doTitulo(hash)
        if (campos.has("nome")) t.put("nome", texto(campos.opt("nome"), 200))
        if (campos.has("descricao")) t.put("descricao", texto(campos.opt("descricao")))
        if (campos.has("etiquetas")) {
            val cru = campos.opt("etiquetas")
            val lista = when (cru) {
                is JSONArray -> (0 until cru.length()).map { cru.optString(it) }
                else -> cru.toString().split(",")
            }
            val limpas = lista.map { texto(it, 40) }.filter { it.isNotEmpty() }.distinct().take(20)
            t.put("etiquetas", JSONArray().apply { limpas.forEach { put(it) } })
        }
        if (campos.has("pasta")) {
            val pasta = campos.optString("pasta").ifEmpty { null }
            t.put("pasta", if (pasta != null && pastas().has(pasta)) pasta else JSONObject.NULL)
        }
        gravar()
        return t
    }

    /** Nome de exibicao de um arquivo (episodio) dentro de um titulo. */
    @Synchronized
    fun editarArquivo(hash: String, caminho: String, nome: String?): JSONObject {
        val t = doTitulo(hash)
        val arquivos = t.optJSONObject("arquivos") ?: JSONObject().also { t.put("arquivos", it) }
        val limpo = texto(nome, 200)
        if (limpo.isNotEmpty()) arquivos.put(caminho, JSONObject().put("nome", limpo))
        else arquivos.remove(caminho)
        gravar()
        return t
    }

    // ------------------------------------------------------------------- capas

    private fun apagarArquivoCapa(nome: String?) {
        if (nome.isNullOrEmpty() || nome == "null") return
        try {
            File(Caminhos.capas(contexto), File(nome).name).delete()
        } catch (e: Exception) {
            // ja nao existe
        }
    }

    private fun extensaoValida(origem: String): String? {
        val caminho = try {
            Uri.parse(origem).path ?: origem
        } catch (e: Exception) {
            origem
        }
        val ext = caminho.substringAfterLast('.', "").lowercase()
        return if (EXT_IMAGEM.contains(ext)) ext else null
    }

    private fun baixarImagem(url: String): Pair<ByteArray, String> {
        cliente.newCall(Request.Builder().url(url).build()).execute().use { r ->
            if (!r.isSuccessful) throw Exception("o servidor respondeu ${r.code}")
            val tipo = (r.header("content-type") ?: "").substringBefore(';').trim()
            if (!tipo.startsWith("image/")) {
                throw Exception("o link nao aponta para uma imagem (${tipo.ifEmpty { "sem tipo" }})")
            }
            val bytes = r.body?.bytes() ?: ByteArray(0)
            if (bytes.size > TAMANHO_MAXIMO_CAPA) throw Exception("a imagem passa de 12 MB")
            val ext = extensaoValida(url)
                ?: MIMES.entries.firstOrNull { it.value == tipo }?.key
                ?: "jpg"
            return Pair(bytes, ext)
        }
    }

    /**
     * Le a imagem que o usuario escolheu. No Android o seletor entrega um
     * content:// (nao um caminho), entao a leitura passa pelo ContentResolver.
     */
    private fun lerImagemLocal(origem: String): Pair<ByteArray, String> {
        val uri = Uri.parse(origem)
        val bytes = if (uri.scheme == "content") {
            contexto.contentResolver.openInputStream(uri)?.use { it.readBytes() }
                ?: throw Exception("nao consegui ler a imagem escolhida")
        } else {
            File(uri.path ?: origem).readBytes()
        }
        if (bytes.size > TAMANHO_MAXIMO_CAPA) throw Exception("a imagem passa de 12 MB")

        val tipo = if (uri.scheme == "content") contexto.contentResolver.getType(uri) ?: "" else ""
        val ext = extensaoValida(origem)
            ?: MIMES.entries.firstOrNull { it.value == tipo.substringBefore(';').trim() }?.key
            ?: "jpg"
        return Pair(bytes, ext)
    }

    /**
     * Guarda a capa dentro dos dados do aplicativo -- uma COPIA, para a
     * biblioteca nao quebrar se o arquivo original for movido ou apagado.
     */
    @Synchronized
    fun definirCapa(alvo: JSONObject, origem: JSONObject): String {
        val tipo = alvo.optString("tipo")
        val id = alvo.optString("id")
        val destino = (if (tipo == "pasta") pastas().optJSONObject(id) else doTitulo(id))
            ?: throw Exception("item não encontrado")

        val url = origem.optString("url")
        val (bytes, ext) = if (url.isNotEmpty()) baixarImagem(url)
        else lerImagemLocal(origem.optString("arquivo"))

        val nome = "${novoId()}.$ext"
        File(Caminhos.capas(contexto), nome).writeBytes(bytes)

        apagarArquivoCapa(destino.optString("capa"))
        destino.put("capa", nome)
        gravar()
        return nome
    }

    @Synchronized
    fun removerCapa(alvo: JSONObject): Boolean {
        val tipo = alvo.optString("tipo")
        val id = alvo.optString("id")
        val destino = if (tipo == "pasta") pastas().optJSONObject(id) else titulos().optJSONObject(id)
        if (destino == null) return false
        apagarArquivoCapa(destino.optString("capa"))
        destino.put("capa", JSONObject.NULL)
        gravar()
        return true
    }

    /** Serve o arquivo da capa para o endereco capa/img/<nome>. */
    fun lerCapa(nome: String?): Pair<ByteArray, String>? {
        val seguro = File(nome ?: "").name
        if (seguro.isEmpty()) return null
        val arquivo = File(Caminhos.capas(contexto), seguro)
        if (!arquivo.exists()) return null
        val mime = MIMES[seguro.substringAfterLast('.', "").lowercase()] ?: "image/jpeg"
        return Pair(arquivo.readBytes(), mime)
    }

    /**
     * O endereco que a interface poe no src da imagem.
     *
     * No desktop era o esquema proprio `capa://img/<nome>`. Aqui a pagina toda
     * e servida de uma origem so (o WebViewAssetLoader), e um caminho relativo
     * dentro dela faz o mesmo papel -- sem precisar registrar esquema nenhum
     * nem afrouxar a CSP.
     */
    fun urlDaCapa(nome: String?): Any {
        if (nome.isNullOrEmpty() || nome == "null") return JSONObject.NULL
        return "capa/img/" + Uri.encode(nome)
    }

    // -------------------------------------------------------------------- juncao

    /** Junta o catalogo da biblioteca com o que o usuario editou. */
    @Synchronized
    fun aplicar(entradas: JSONArray): JSONArray {
        val saida = JSONArray()
        for (i in 0 until entradas.length()) {
            val e = JSONObject(entradas.optJSONObject(i).toString())
            val hash = e.optString("hash")
            val meta = titulos().optJSONObject(hash)

            e.put("nomeOriginal", e.optString("nome"))

            if (meta == null) {
                e.put("etiquetas", JSONArray())
                e.put("pasta", JSONObject.NULL)
                e.put("capa", JSONObject.NULL)
                saida.put(e)
                continue
            }

            val nome = meta.optString("nome")
            if (nome.isNotEmpty()) e.put("nome", nome)
            e.put("descricao", meta.optString("descricao"))
            e.put("etiquetas", meta.optJSONArray("etiquetas") ?: JSONArray())
            e.put("pasta", meta.opt("pasta") ?: JSONObject.NULL)
            e.put("capa", urlDaCapa(meta.optString("capa")))

            val nomesDeArquivo = meta.optJSONObject("arquivos") ?: JSONObject()
            val arquivos = e.optJSONArray("arquivos") ?: JSONArray()
            val novos = JSONArray()
            for (j in 0 until arquivos.length()) {
                val a = JSONObject(arquivos.optJSONObject(j).toString())
                val original = a.optString("nome")
                a.put("nomeOriginal", original)
                val escolhido = nomesDeArquivo.optJSONObject(a.optString("caminho"))?.optString("nome")
                if (!escolhido.isNullOrEmpty()) a.put("nome", escolhido)
                novos.put(a)
            }
            e.put("arquivos", novos)
            saida.put(e)
        }
        return saida
    }

    fun pastasParaInterface(): JSONArray {
        val saida = JSONArray()
        for (p in listarPastas()) {
            saida.put(JSONObject(p.toString()).put("capa", urlDaCapa(p.optString("capa"))))
        }
        return saida
    }
}
